prboom2/src/dsda/utility/string_view.h

```diff
@@ -0,0 +1,58 @@
+//
+// Copyright(C) 2023 by Ryan Krafnick
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// DESCRIPTION:
+//	DSDA Utility String View
+//
+
+#ifndef __DSDA_UTILITY_STRING_VIEW__
+#define __DSDA_UTILITY_STRING_VIEW__
+
+#include "doomtype.h"
+
+// Non-allocated, read-only view into a string, like C++ string_view
+typedef struct {
+  const char* string;
+  size_t size;
+} dsda_string_view_t;
+
+void dsda_InitStringView(dsda_string_view_t* sv, const char* string, size_t size);
+
+dboolean dsda_IsStringViewEmpty(const dsda_string_view_t* sv);
+
+void dsda_StringViewAtOffset(const dsda_string_view_t* sv, size_t offset, dsda_string_view_t* ofs);
+
+// Splits `sv` at the first instance of `c`, putting everything up to and including it in `before`
+// and everything after in `after`.  `before` or `after` may safely alias `sv`.  If `c` does not
+// occur in `sv`, `before` is set to `sv` and `after` is set to empty.  Returns `true` if
+// an occurence was found.
+dboolean dsda_SplitStringViewAfterChar(const dsda_string_view_t* sv, char c,
+                                       dsda_string_view_t* before,
+                                       dsda_string_view_t* after);
+
+// Like the above, except `c` is included in `after` if found
+dboolean dsda_SplitStringViewBeforeChar(const dsda_string_view_t* sv, char c,
+                                        dsda_string_view_t* before,
+                                        dsda_string_view_t* after);
+
+// Advances `sv` past the next line and sets `line` to it, including any trailing '\r'
+// or '\n'.  If the string does not end with a '\n`, `line` is set to `sv` and `sv`
+// is emptied.  Returns `false` if there was no more data, i.e. `sv` was empty.
+dboolean dsda_AdvanceStringViewToNextLine(dsda_string_view_t* sv, dsda_string_view_t* line);
+
+dboolean dsda_StringViewStartsWithPrefix(const dsda_string_view_t* sv, const char* prefix);
+
+void dsda_StringViewAfterChars(const dsda_string_view_t* sv, const char* chars,
```

Comment:
The comment for this said it removed chars from sv, but it doesn't appear to do that - it just sets the after to after the chars. Is it supposed to?

Reply:
It should "return" a new string, since there's no reason to require that `sv` be modified in place, and you can just use it as the output if that's what you want.

This function should really be called "TrimLeft" or "StripBefore" or something, as it's intended to work like [`lstrip`](https://docs.python.org/3/library/stdtypes.html#str.lstrip) in python.  There would be a corresponding function to strip from the right side of the string, and one that does both at once.  It's usually used to discard leading/trailing whitespace.

Reply:
It would confuse me to use that name and have it not mutate or return newly allocated data, that's why I've changed the names like this.

Reply:
What would you call `rstrip`, which strips from the right side of the string, or `strip`, which does both?

Reply:
dsda_StringViewBeforeChars I guess :^)

Reply:
By the way, something for you to keep in mind in the future. If you are going to do something that you can encapsulate and decouple cleanly, where you would benefit from the c++ stuff, it's completely fine with me if you do it in c++. We have this for the udmf and umapinfo parsers, and preprocessing text is probably another thing that can be cleanly isolated.